controllers/utils/failuredomains.go

```diff
@@ -0,0 +1,163 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	infrav1 "sigs.k8s.io/cluster-api-provider-cloudstack/api/v1beta2"
+	"sigs.k8s.io/cluster-api-provider-cloudstack/pkg/cloud"
+
+	"github.com/pkg/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// CreateFailureDomain creates a specified CloudStackFailureDomain CRD owned by the ReconcilationSubject.
+func (r *ReconciliationRunner) CreateFailureDomain(fdSpec infrav1.CloudStackFailureDomainSpec) error {
+	csFD := &infrav1.CloudStackFailureDomain{
+		ObjectMeta: r.NewChildObjectMeta(fdSpec.Name),
+		Spec:       fdSpec,
+	}
+	return errors.Wrap(r.K8sClient.Create(r.RequestCtx, csFD), "creating CloudStackFailureDomain")
+}
+
+// CreateFailureDomains creates a CloudStackFailureDomain CRD for each of the ReconcilationSubject's FailureDomains.
+func (r *ReconciliationRunner) CreateFailureDomains(fdSpecs []infrav1.CloudStackFailureDomainSpec) CloudStackReconcilerMethod {
+	return func() (ctrl.Result, error) {
+		for _, fdSpec := range fdSpecs {
+			if !strings.HasSuffix(fdSpec.Name, "-"+r.CAPICluster.Name) { // Add cluster name suffix if missing.
+				fdSpec.Name = fdSpec.Name + "-" + r.CAPICluster.Name
+			}
+			if err := r.CreateFailureDomain(fdSpec); err != nil {
+				if !strings.Contains(strings.ToLower(err.Error()), "already exists") {
+					return reconcile.Result{}, errors.Wrap(err, "creating CloudStackFailureDomains")
+				}
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+}
+
+// GetFailureDomains gets CloudStackFailureDomains owned by a CloudStackCluster.
+func (r *ReconciliationRunner) GetFailureDomains(fds *infrav1.CloudStackFailureDomainList) CloudStackReconcilerMethod {
+	return func() (ctrl.Result, error) {
+		capiClusterLabel := map[string]string{clusterv1.ClusterLabelName: r.CSCluster.GetLabels()[clusterv1.ClusterLabelName]}
+		if err := r.K8sClient.List(
+			r.RequestCtx,
+			fds,
+			client.InNamespace(r.Request.Namespace),
+			client.MatchingLabels(capiClusterLabel),
+		); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to list failure domains")
+		}
+		return ctrl.Result{}, nil
+	}
+}
+
+// RemoveExtraneousFailureDomains deletes failure domains no longer listed under the CloudStackCluster's spec.
+func (r *ReconciliationRunner) RemoveExtraneousFailureDomains(fds *infrav1.CloudStackFailureDomainList) CloudStackReconcilerMethod {
+	return func() (ctrl.Result, error) {
+		// Toss together a precense map.
+		fdPresenceByName := map[string]bool{}
+		for _, fdSpec := range r.CSCluster.Spec.FailureDomains {
+			name := fdSpec.Name
+			if !strings.HasSuffix(name, "-"+r.CAPICluster.ClusterName) { // Add cluster name suffix if missing.
```

Comment:
We are comparing suffix in if using r.CAPICluster.ClusterName, but set value using r.CAPICluster.Name

Actually, r.CAPICluster.ClusterName has empty string while r.CAPICluster.Name has expected value of cluster name.

According to ref: https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1@v0.23.0#ObjectMeta.ClusterName, clusterName should not be used yet.

I think this is the reason in my test that all VMs are deleted while upgrading. 

We also need to scan our code base for using of clusterName.